src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,256 @@
+    mutex.try_lock().unwrap() // panic here if in use
+}
+
+/// Driver output [Relay] type that controlls the set of two relays that is present on both of
+/// the driver output channels.
+/// [Relay] contains a mutex of a shared MCP23008, an I2C io expander chip with 8 GPIOs.
+/// 4 of those GPIOs are used to controll one set of relays (aka one [Relay]):
+///     - `k1_en_n` and `k1_en` differentially control the state of K1
+///     - `k0_d` is that data input of a flipflop which controls the state of K0
+///     - `k0_cp` is the clock input if the flipflop. Note that there is additional circuitry
+///        which ensures this clock input cannot be driven if K1 is not in the correct state.
+///        See Driver schematic for exact details.
+/// A `delay` is used to generate a rising edge for the flipflip with enaugh timing margins.
+pub struct Relay<'a, I2C: WriteRead + Write> {
+    mutex: &'a spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+    delay: asm_delay::AsmDelay,
+    k1_en_n: RelayPin,
+    k1_en: RelayPin,
+    k0_d: RelayPin,
+    k0_cp: RelayPin,
+}
+
+impl<'a, I2C, E> Relay<'a, I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    const K0_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10);
+    const K1_DELAY: fugit::MillisDuration<u64> =
+        fugit::MillisDurationU64::millis(10);
+
+    /// Construct a new [Relay].
+    ///
+    /// # Args
+    /// * `mutex`   - mutex of a shared MCP23008
+    /// * `ccdr`    - core clocks to construct a `delay`
+    /// * `channel` - Driver channel to construct the [Relay] for
+    pub fn new(
+        mutex: &'a spin::Mutex<Mcp230xx<I2C, Mcp23008>>,
+        ccdr: &rcc::CoreClocks,
+        channel: Channel,
+    ) -> Self {
+        let (k1_en_n, k1_en, k0_d, k0_cp) = if channel == Channel::LowNoise {
+            (
+                RelayPin::LN_K1_EN_N,
+                RelayPin::LN_K1_EN,
+                RelayPin::LN_K0_D,
+                RelayPin::LN_K0_CP,
+            )
+        } else {
+            (
+                RelayPin::HP_K1_EN_N,
+                RelayPin::HP_K1_EN,
+                RelayPin::HP_K0_D,
+                RelayPin::HP_K0_CP,
+            )
+        };
+        let mut mcp = get_mcp(mutex);
+        // set GPIOs to default position
+        mcp.set_gpio(k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(k1_en_n.into(), Level::High).unwrap();
+        mcp.set_gpio(k0_d.into(), Level::Low).unwrap();
+        mcp.set_gpio(k0_cp.into(), Level::Low).unwrap();
+
+        let delay = asm_delay::AsmDelay::new(asm_delay::bitrate::Hertz(
+            ccdr.c_ck().to_Hz(),
+        ));
+        Relay {
+            mutex,
+            delay,
+            k1_en_n,
+            k1_en,
+            k0_d,
+            k0_cp,
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            *Disabled + Enable / engage_k0 = EnableWaitK0,
```

Comment:
Comment that the hardware guarantees that initially you are in `Disabled`.

Also review whether there is a way for the hardware to go out of sync (`TEMP`?) with your internal state machine state.